### PR TITLE
Tagchanges

### PR DESCRIPTION
**Pull request by skyfrench**
Modifications to dls_ade to enable dls_release.py to tag a commit, prior to release.

New argument added '-s' (--sha)
e.g. 
``
dls-release.py <module> <tag> -m "user message" --sha <commitID>
``
- If a --sha argument is not given - the script will proceed as before provided the tag ('release') already exists. (If the tag does not conform to the convention, the script will produce a warning - but still go ahead as before.)

- If a --sha argument is given, and tag is of the permitted format, and does not already exist - the script will first apply the tag at the commit ID before proceeding. The script will abort if the tag already exists.
 
(New script added) tags must be of the form: X-Y[-Z][dlsA[-B]], where X, Y, Z, A & B are integers.

The commit message upon tagging is of the form:
DLS Release {tag}: {user message}

### CODE REVIEW

File: dls_ade/tag_utility.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+    """
+
+    check = re.compile('[0-9]+\-[0-9]+(\-[0-9]+)?(dls[0-9]+(\-[0-9]+)?)?')
+    result = check.match(tag)
+
+    if result is None:
+        raise ParsingError("Invalid tag " + tag)
+
+    if result.group() != tag:
```
Comment (GDYendell) [inline dls_ade/tag_utility.py]: if result is None or result.group() != tag:
￼        raise ParsingError("Invalid tag " + tag)

File: dls_ade/dls_release.py [hunk 2]
```diff
@@ -410,14 +418,58 @@ def _main():
...
         vcs.set_branch(args.branch)
 
+    releases = vcs.list_releases()
+
+    tag = args.release
+    tag_is_valid = True
+    try:
+        check_tag_is_valid(tag)
+    except ParsingError as err:
+        tag_is_valid = False
```
Comment (GDYendell) [inline dls_ade/dls_release.py]: if you make the method return True/False you can just do:

tag_is_valid = check_tag_is_valid()

File: dls_ade/vcs_git.py [hunk 3]
```diff
@@ -368,18 +368,19 @@ def set_version(self, version):
             raise VCSGitError('Version \'{}\' does not exist in tag list: {}'.format(version, self.list_releases()))
         self._version = version
 
-    def push_to_remote(self, remote_name="origin", branch_name="master"):
+    def push_to_remote(self, remote_name="origin", branch_or_tag_name="master"):
```
Comment (GDYendell) [inline dls_ade/vcs_git.py]: The generic term in git for a branch / tag / hash is 'ref'

**Comment by skyfrench** [comment]
Thanks for the review - changes made.

**Comment by skyfrench** [comment]
Now rebased and with a tidy up of the dls_utilities imports.

**Comment by skyfrench** [comment]
Modifications have been made to incorporate the needs of [CGP-90](https://jira.diamond.ac.uk/browse/CGP-90) - enabling test on build server without a git tag.

If the release positional argument is missing, but the -T flag is used and --commit / -c set - the test build will checkout and build the module at the commit.

Modifications were required to the scripts in Linux/*.sh as discussed in [CGP-90](https://jira.diamond.ac.uk/browse/CGP-90) (git clone, then git checkout - to replace the previous clone, fetch depth=1 then checkout commands):
- I have tested with example commits in dls_ade, using the python.sh script
- The git commands worked for the test cases - but I'm new to git, so there may be a flaw in the modifications to the commands that I'm not aware of?
- Do the Windows scripts still work / need modifying?

**Comment by skyfrench** [comment]
Modifications to:

- Allow a clone all iff running at test build; this will allow git checkout SHA to find the relevant point
- Restore production build behaviour to previous behaviour (clone depth=100, follow by fetch depth=1)

**Comment by ulrikpedersen** [comment]
> Do the Windows scripts still work / need modifying?

Yes, good point: don't forget the windows scripts, please. These are almost more important to enable test-builds for as Windows builds are often quite a bit more difficult than Linux builds...

**Comment by skyfrench** [comment]
Windows scripts have now been updated.

**Comment by skyfrench** [comment]
@gilesknap worth giving this a test with your current example of where this would be useful?

**Comment by skyfrench** [comment]
Branch has been rebased.

**Pull request merged**